### DISCUSSION TITLE
gd扩展的支持

### DISCUSSION THREAD

**Discussion by liida** [dixyes/lwmbs#4] [Ideas]
可以增加gd库扩展的支持吗

**Comment by liida** [comment]
特别是windows下

**Comment by dixyes** [comment]
我倒是早就想搞了，但gd依赖了n个图像处理库，它们的构建支持，许可证都得处理。
而且它会让现在的ci构建变得更加臃肿（预期增加5M+大小），所以有空了可能会做支持，但不太可能放到ci里

**Reply by jingjingxyk** [reply]
这个其实我是搞定了GD 。只是迁移到  lwmbs   ，需要添加  pkg-config 

比如freetype 库 
```shell


            BZIP2_CFLAGS="-I{$bzip2_prefix}/include"  \
            BZIP2_LIBS="-L{$bzip2_prefix}/lib -lbz2"  \
            CPPFLAGS="$(pkg-config --cflags-only-I --static zlib libpng  libbrotlicommon  libbrotlidec  libbrotlienc)" \
            LDFLAGS="$(pkg-config  --libs-only-L   --static zlib libpng  libbrotlicommon  libbrotlidec  libbrotlienc)" \
            LIBS="$(pkg-config     --libs-only-l   --static zlib libpng  libbrotlicommon  libbrotlidec  libbrotlienc)" \
            ./configure --prefix={$freetype_prefix} \
            --enable-static \
            --disable-shared \
            --with-zlib=yes \
            --with-bzip2=yes \
            --with-png=yes \
            --with-harfbuzz=no  \
            --with-brotli=yes \
            --enable-freetype-config


```


如果没有pkg-config ，你看上面这一步怎么处理才好?

**Comment by try-to** [comment]
+1

**Comment by dixyes** [comment]
才想起来gd（一年前）已经支持了...

~~因为原神长草了~~

总记得这些图像影像库全是gpl 但整理了下相关库 应该是可以支持

| 库 | 支持的格式 | 许可证限制 | 大小（大约） | 必选依赖 | 可选依赖 |
| - | - | - | - | - | - |
| GD本身 | gif bmp png(通过必须依赖libpng) | PHP许可证，可以静态发布 | 100k | libpng | |
| libpng | png | libpng许可证 可以静态发布 | 200k | zlib |  |
| libavif | avif(对于编码的支持需要额外依赖) | BSD 可以静态发布 | 150k | libyuv | libdav1d libgav1 librav1e svt-av1 libaom |
| libyuv | - | BSD 可以静态发布 | 50k | | libjpeg |
| libwebp | webp | BSD 可以静态发布 | 400k | | |
| libjpeg | jpeg | ijg许可证 可以静态发布 | 700k | | |
| libjpeg-turbo | jpeg | ijg+bsd+zlib许可证 可以静态发布 | 700k | | |
| libxpm | xpm | X11许可证 可以静态发布 但做不到（虽然许可证支持，但x11本身不支持静态连接/使用） | ∞ | 一堆 | |
| libfreetype | ttf/字体支持 | freetype/GPL许可证 可以静态发布 | 800k | 需要目标操作系统上一些配置文件 | libpng zlib libbz2 |

max flavor里没有开启gd，那么来选择下max flavor支持的库吧：

1. libgd+libpng 支持bmp png gif 二进制+300k
2. libgd+libpng+libjpeg 支持bmp png gif jpg 二进制+1M
3. 能搞的全都要 二进制+2M起

@try-to @jingjingxyk

**Reply by jingjingxyk** [reply]
@dixyes  能支持都支持，。

GD 依赖的的静态链接库，除了libXpm 没有搞定，其它我都搞定了（详见swoole-cli build_native_php 分支）。现在的问题是本项目暂时没有 pkg-config 支持不太好弄。



swoole-cli  项目里  [`build_native_php` ](https://github.com/swoole/swoole-cli/tree/build_native_php)  、[`experiment` ](https://github.com/swoole/swoole-cli/tree/experiment)里面的GD 库支持我已经完成了

构建结果详见：[ experiment](https://github.com/swoole/swoole-cli/actions/runs/5904952818/job/16018096900#step:7:89)

[static-php-cli 我也折腾过](https://github.com/crazywhalecc/static-php-cli/pull/40/files#diff-a29be5d60a878c93571626cbdb3861b4680f7e48503f6d00733f9afe843f107a)，因没有pkg-config 支持 ，不好弄。
我等着这个[static-php-cli  支持 pkg-config 的 PR合并](https://github.com/crazywhalecc/static-php-cli/pull/130)合并以后，就支持pkg-config  了,代码搬过来就能用

**Reply by dixyes** [reply]
不存在的 我已经在尽量抛弃pkgconfig 全搞cmake了 况且lwmbs不是这样工作的 lwmbs有自己的抽象 这个抽象不关心如何实现 你只要实现相关库的构建就行

**Reply by jingjingxyk** [reply]
不是每个库都能用 cmake 来构建的。 这里列出三种： 常见三种构建方式： https://github.com/swoole/swoole-cli/blob/build_native_php/sapi/src/builder/library/aaa_example.php

另外很多库，虽然使用cmake ，但是cmake 构建过程中仍然会使用pkgconfig 查找包 详见：libzip https://github.com/swoole/swoole-cli/blob/build_native_php/sapi/src/builder/library/libzip.php

**Reply by dixyes** [reply]
不是 我不关心你构建库用了什么工具 你用ninja gyp都可以 我真不关系

**Comment by jingjingxyk** [comment]
把pkg-config 支持加上， 代码搬过来就能用

依赖的静态链接库构建过程
[libaom](https://github.com/swoole/swoole-cli/blob/build_native_php/sapi/src/builder/library/aom.php)
[dav1d](https://github.com/swoole/swoole-cli/blob/build_native_php/sapi/src/builder/library/dav1d.php)
[libavif](https://github.com/swoole/swoole-cli/blob/build_native_php/sapi/src/builder/library/libavif.php)
[SVT-AV1](https://github.com/swoole/swoole-cli/blob/build_native_php/sapi/src/builder/library/svt_av1.php)
[libwebp](https://github.com/swoole/swoole-cli/blob/build_native_php/sapi/src/builder/library/libwebp.php)

**Comment by dixyes** [comment]
基本完成了 gd png jpg webp 懒狗懒得做avif了 反正我没见人用过这格式
明天的自动构建应该会包含了

**Reply by jingjingxyk** [reply]
avif 图片格式应用案例： 
https://www.ixigua.com/

![image](https://github.com/dixyes/lwmbs/assets/6836228/b9a28187-3dd3-4d72-a25e-1bdb0ceaa8a9)
